fate test for dnn_module v3

Avery-xu opened the pull request.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d.c:
```diff
@@ -152,35 +211,35 @@ static void * dnn_execute_layer_conv2d_thread(void *threadarg)
-                                                   input[y_pos * src_linesize + x_pos * conv_params->input_num + ch];
-                            }
-
+                    output[n_filter] += conv_params->biases[n_filter];
```
why we don't put this into _c function.

Avery-xu replied:
This should be put into _c function and can be used as initial value of output.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d.c:
```diff
@@ -223,6 +282,7 @@ int dnn_execute_layer_conv2d(DnnOperand *operands, const int32_t *input_operand_
     thread_common_param.output_operand_index = output_operand_index;
     thread_common_param.parameters = parameters;
     thread_common_param.ctx = ctx;
+
```
don't add extra empty line

guoyejun reviewed (commented):
-cpuflags flags (global)
Allows setting and clearing cpu flags. This option is intended for testing. Do not use it unless you know what you’re doing.

don't add -cpuflags 0x100 in the commit log since it is for testing.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d.c:
```diff
@@ -233,7 +234,10 @@ static void * dnn_execute_layer_conv2d_thread(void *threadarg)
+    if ((thread_common_param->channel_step >= 8) && (conv_params->input_num >= 8)) {
+        ff_dnn_execute_layer_conv2d_avx2(&exe_param);
+    }
+    else if ((thread_common_param->channel_step >= 4) && (conv_params->input_num >= 4)) {
```
if () {

} else if () {

} else {

}

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d.c:
```diff
@@ -296,6 +300,8 @@ int dnn_execute_layer_conv2d(DnnOperand *operands, const int32_t *input_operand_
         int cpu_flags = av_get_cpu_flags();
         if (EXTERNAL_SSE4(cpu_flags))
             thread_common_param.channel_step = 4;
```
what does channel mean for cpu flag?

a better naming might be asm_type or something else, and do the judgement here, like below:

if ((EXTERNAL_SSE4(cpu_flags) && (conv_params->input_num >= 4))
   thread_common_param.asm_type = SSE.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm:
```diff
@@ -0,0 +1,214 @@
+INIT_XMM sse4
+cglobal dnn_execute_layer_conv2d, 8, 15, 3, exe_param,\
+    x, y, n_filter, cha, kernel_x, kernel_y, x_pos, y_pos, kernel_pos,\
+    input, output, kernel, tmp1, tmp2
```
%macro cglobal 1-2+ "" ; name, [PROLOGUE args]

why we need to add x, y, n_filter, ... which are not function args, they are also not used in the below code.

Avery-xu replied:
x, y, n_filter were names of registers, which can be seen as variables.

guoyejun replied:
but there variables are not used in the below code.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm:
```diff
@@ -0,0 +1,214 @@
+    mov outputq, [exe_paramq + 14 * 4 + 8]
+    mov kernelq, [exe_paramq + 14 * 4 + 2 * 8]
+
+    mov yd, thread_start
```
what do the suffix 'q' and 'd' mean?

Avery-xu replied:
d, q were data length of the registers. d means the lower 32-bit of a 64-bit register, and q means the whole register.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm:
```diff
@@ -0,0 +1,214 @@
+        .loop_filter:
+            xorps m2, m2
+            xor kernel_yd, kernel_yd
+
```
why empty line here and there? 

better to add comments for a different logic block.

guoyejun commented on libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm:
```diff
@@ -0,0 +1,214 @@
+            xor kernel_yd, kernel_yd
+
+            mov tmp1d, kernel_yd
+            sub tmp1d, radius
```
possible to have a better naming for m0, m1, m2, tmp1d and tmp2d ...?

Avery-xu replied:
m0, m1,..., were vector registers. In XMM case, they are 128-bit width. Usually we don't rename them. As for tmp1 and tmp2, I will get a new name for them.

guoyejun replied:
thanks, and also add comments if necessary to help reviewers to understand easily.

Avery-xu committed:
dnn_backend_native_layer_conv2d.c: fix memory allocation bug in multithread function.

Before patch, memory was allocated in each thread functions,
which may cause more than one time of memory allocation and
cause crash.

After patch, memory is allocated in the main thread once,
an index was parsed into thread functions. Bug fixed.

Signed-off-by: Xu Jun <xujunzz@sjtu.edu.cn>

Avery-xu committed:
dnn_backend_native_layer_conv2d.c: refine code.

Move thread area allocate out of thread function into
main thread.

Signed-off-by: Xu Jun <xujunzz@sjtu.edu.cn>
